### PR TITLE
MM-11359: support excluding results from search

### PR DESCRIPTION
**Pull request by liusy182**
#### Summary

Back end work for https://github.com/mattermost/mattermost-server/issues/11071

MySQL and PostgreSQL are tested working. 
Not able to test / work on ElasticSearch.

#### Ticket Link

https://github.com/mattermost/mattermost-server/issues/11071

### CODE REVIEW

**Comment by grundleborg** [comment]
@wiersgallak please could you give this a PM review?

**Comment by wiersgallak** [comment]
@liusy182 I am not getting the expected results when adding the "-" to exclude dates using the "on" "after" or "before" modifiers.  

I did confirm this is as expected for "from" modifier and using quotes for the search term.

**Comment by liusy182** [comment]
@wiersgallak I should have mentioned that the implementation did not cover `on`, `before` and `after`. It probably make sense to include `on`, but `before` and `after` is already supported by modifying query slightly. e.g. "something -before:2019-01-10" is equivalent to "something after:2019:01-10". Should `on` be covered? I can make the change if it is part of the plan.

**Comment by wiersgallak** [comment]
@liusy182 Yes, please add support for "on".  When searching for a keyword that was used heavily on a specified date, it would be helpful for users to be able to omit those results from their query so they can narrow in on posts without having to navigate many pages of results. 

I can see a use case for supporting ``after`` and ``before`` when they are used together.  For instance, I want to find posts that included a key word outside of a time frame, so even though we do have the logic because of the inverse nature of ``after`` and ``before`` supporting both of these modifiers provides the ability to search outside of the timeframe.  This is a similar experience to ``on`` in that it allows the user to narrow in on posts without having to navigate sets of results if they already know they are not relevant within that specified time.

**Comment by liusy182** [comment]
@wiersgallak got it totally make sense. let me update the PR to include `on`, `before` and `after` exclusion support.

**Comment by liusy182** [comment]
@wiersgallak updated with `on`, `before` and `after` added ^^

**Comment by liusy182** [comment]
have you got time to check on latest commit @wiersgallak ?

**Comment by wiersgallak** [comment]
@liusy182 thanks for adding the modifier to the ``on``, ``before`` and ``after``  - these are working as I would expect.  Can you please fix the checks and we will get the developer's review completed.

**Comment by liusy182** [comment]
ci is broken again due to diverged files, am i suppose to merge again for review?

**Comment by grundleborg** [comment]
@liusy182 don't worry about it - once we get the dev review completed you can merge again but for now there's no need.

**Comment by liusy182** [comment]
cool thanks @grundleborg this has been idle for a while, is there anyone working on the review?

File: model/search_params_test.go [hunk 1]
```diff
@@ -59,250 +64,846 @@ func TestSplitWords(t *testing.T) {
...
-		t.Fatalf("got incorrect flags %v", flags)
 	}
+}
 
-	if words, flags := parseSearchFlags(splitWords("test after:2018-7-1")); len(words) != 1 || words[0] != "test" {
-		t.Fatalf("got incorrect words %v", words)
-	} else if len(flags) != 1 || flags[0][0] != "after" || flags[0][1] != "2018-7-1" {
-		t.Fatalf("got incorrect flags %v", flags)
-	}
+func TestParseSearchFlags(t *testing.T) {
```
Comment (mgdelacroix) [inline model/search_params_test.go]: These are great candidates for table testing

File: model/search_params.go [hunk 2]
```diff
@@ -79,9 +132,13 @@ func splitWords(text string) []string {
 				foundQuote = false
 				location = i + 1
 			} else {
-				words = append(words, strings.Fields(text[location:i])...)
+				nextStart := i
+				if i-1 >= 0 && text[i-1] == '-' {
```
[the remark this answers is not in the document]
Comment (liusy182) [reply model/search_params.go]: sounds good will update this.

File: store/sqlstore/post_store.go [hunk 3]
```diff
@@ -721,34 +721,154 @@ var specialSearchChar = []string{
 	":",
 }
 
+func (s *SqlPostStore) buildSearchChannelFilterClause(channels []string, paramPrefix string, exclusion bool, queryParams map[string]interface{}) (string, map[string]interface{}) {
+	if len(channels) > 1 {
+		clause := ":" + paramPrefix + "0"
```
[the remark this answers is not in the document]
Comment (liusy182) [reply store/sqlstore/post_store.go]: sounds good will update both places

File: store/sqlstore/post_store.go [hunk 4]
```diff
@@ -721,34 +721,154 @@ var specialSearchChar = []string{
...
+
+	if len(params.ExcludedDate) > 0 {
+		excludedDateStart, excludedDateEnd := params.GetExcludedDateMillis()
+		queryParams["ExcludedDateStart"] = strconv.FormatInt(excludedDateStart, 10)
+		queryParams["ExcludedDateEnd"] = strconv.FormatInt(excludedDateEnd, 10)
+
+		searchQuery += "AND CreateAt NOT BETWEEN :ExcludedDateStart AND :ExcludedDateEnd "
+	}
+
+	if len(params.AfterDate) > 0 {
```
[the remark this answers is not in the document]
Comment (liusy182) [reply store/sqlstore/post_store.go]: sure i will wait for @wiersgallak to comment first before making changes.

[the remark this answers is not in the document]
Comment (wiersgallak) [reply store/sqlstore/post_store.go]: > * `on:2019-07-19 -on:2019-01-01` would be the same as `on:2019-07-19`

Agreed. 

> * `-on:2019-07-19 before:2019-01-01` would be the same as `-on:2019-07-19`
 
I would expect this return posts "not on 2019-07-19 and "before  2019-01-01"

for another example a bit different to illustrate my expectations with using `on` with `before` and `after`: 
* `-on: 2019-07-19 after: 2019-07-01` would return posts any date after 2019-07-01 excluding (or except) posts from 2019-07-19.  

Please let me know if you have questions on that last example.

**Comment by liusy182** [comment]
@mgdelacroix fixed review comments. 

Apologize for the forced push if it messes up diffs. i was too used to rebase and it didn't go well.

**Comment by srkgupta** [comment]
This will be tested as part of the 5.16 release & epic `MM-11354: Advanced Search & Search Improvements`. Removing the QA Review label.

**Pull request merged**